index.js

```diff
@@ -5,18 +5,44 @@
  */
 
 const rescue = f => source => (start, sink) => {
-  start === 0 && source(start, (t, d) => {
-    if (t === 2 && d != null) {
-      try {
-        const newSource = f(d);
-        newSource(start, sink);
-      } catch (e) {
-        sink(2, e);
+  if (start !== 0) return;
+
+  let rescued = false;
+  let sourceTalkback;
+
+  const talkback = (t, d) => {
+    sourceTalkback(t, d);
+  }
+
+  const innerSink = (t, d) => {
+    if (t === 0) {
+      sourceTalkback = d;
+
+      if (rescued) {
+        talkback(1);
+        return;
       }
-    } else {
+
+      sink(t, talkback);
+      return;
+    }
+
+    if (t !== 2 || !d || rescued) {
```

Comment:
use `d !== undefined` instead of `!d` :-) `d` can be an empty string, 0, null, etc

Reply:
> A callbag is terminated when the first argument is 2 and the second argument is either undefined (signalling termination due to success) or any truthy value (signalling termination due to failure).

This part of the spec is kinda non exhaustive - we have to assume in JS either first part (undefined as success) or truthy (failure). 

Most implementations I've seen in the wild go with second one, checking for truthiness instead of the strict comparison to undefined.